---
output: html_document
title: "genotype comparison plots"
---

\n

This R markdown file compares genotype determinations when phasing was performed by ancestry vs. all participants.

* rs713598_G, rs1726866_G, rs10246949_C = dosages coded as 0/1/2 - from plink2 \
* snp_0, snp_1 = alleles for parent _0 & _1 coded as C/G (rs713598), A/G (rs1726866) or C/T (rs10246939) - from phasing with shapeit2 \
* haplo_0, haplo_1 = haplotypes for parent 0 and 1 coded as CAT/GGC/etc. - from hapotype phasing with shapeit2 \
* diplo_0, diplo_1 = diplotypes from haplo_0/haplo_1 coded as CAT/CAT // GGC/GGC // etc. - from combining haplotypes in R \
* taster_status = taster status based on diplotypes coded as Nontaster (CAT/CAT) // Taster (CAT/GGC or GGC/CAT) // Supertaster (GGC/GGC) \ // Other (XXX/XXX) \


```{r setup, include=F}
#command args
args = commandArgs(trailingOnly = TRUE)
ANC = "EUR" #args[1]

knitr::opts_chunk$set(echo = FALSE, message = FALSE,warning = FALSE, fig.path = paste0("../output/haplos_summary/", ANC, "_"))
suppressMessages(silent <- lapply(
  c("tidyverse", "data.table", "knitr", "kableExtra", "paletteer", "gridExtra", "ggpubr"), 
  library, character.only = TRUE))

palettes <- list(NatComms=paletteer_d("ggsci::nrc_npg", n=10),
                 HeatMapOr=paletteer_d("ggsci::orange_material", n=10),
                 HeatMapYlRd=rev(paletteer::paletteer_c("grDevices::heat.colors", n=100)))

```

```{r inputs}
ANCs <- c("ALL", ANC)
ANCs.l <- as.list(ANCs)

genotypes <- c("rs713598_G", "rs713598_0", "rs713598_1", "rs1726866_G", "rs1726866_0", "rs1726866_1", "rs10246939_C", "rs10246939_0", "rs10246939_1", "haplo_0", "haplo_1", "diplo", "taster_status")
bitter_traits <- c("addsalt_freq_QT", "coffee", "tea")
gPCs <- c(paste0("gPC", 1:10))

# load basic functions
source("basic_functions.R")

# load data for ALL & ANC
df <- lapply(ANCs, function(x) {readRDS(paste0("../data/processed/ukb_analysis_", x, "_rm5sd.rda")) %>%
                select(id, all_of(genotypes), all_of(bitter_traits), all_of(gPCs)) %>% 
    rename_with(~str_c(., ".", x), .cols=-"id")} ) %>%
  Reduce(function(df1, df2) right_join(df1, df2, by = "id"), .)
```

```{r summarise-diplos}
# group by extremely rare(lt 0.01%), rare (0.01 to 1%), common (gt 1%)
diplo_gt05_ALL <- names(which(prop.table(table(df %>% select("diplo.ALL"))) >0.005))
diplo_gt05_ANC <- names(which(prop.table(table(df %>% select("diplo.ALL"))) >0.005))
haplo_gt05_ALL <- names(which(prop.table(table(df %>% select(starts_with("haplo_0") & ends_with("ALL")))) >0.005))
haplo_gt05_ANC <- names(which(prop.table(table(df %>% select(starts_with("haplo_0") & ends_with(ANC)))) >0.005))

ids_gt01 <- #
  df %>% select(id, starts_with("diplo") | starts_with("haplo_0")) %>% 
  rename_with( ~gsub(ANC, "ANC", .)) %>%
  filter(diplo.ALL %in% diplo_gt05_ALL & diplo.ANC %in% diplo_gt05_ANC) %>% 
  filter(haplo_0.ALL %in% haplo_gt05_ALL & haplo_0.ANC %in% haplo_gt05_ANC) %>% 
  select(id)

df <- df %>% filter(id %in% ids_gt01$id)
```

```{r geno-compare-fns}
## Build functions
# genotype match table
geno_match <- function(g) {
  df %>% select(id, starts_with(g)) %>%  
    rename_with(~gsub(".*ALL", "ALL", gsub(paste0(".*",ANC), "ANC", .))) %>%
    mutate(match=ifelse(ANC == ALL, 1, 0)) %>% group_by(ANC) %>%
    summarise(Agree=sum(match==1, na.rm=T),
              Disagree=sum(match==0, na.rm=T)) %>% 
    filter(!is.na(ANC)) #%>%
    # Delete genotypes in <0.01 % of both populations
   # filter(!(Agree < nrow(df)*0.0001 & Disagree < nrow(df)*0.0001))
}

# genotype match barplot
plot_geno_match <- function(g) {
  yscale <- max(geno_match(g)[,2:3])*1.10
  geno_match(g) %>% pivot_longer(-ANC) %>% 
    ggplot(aes(x=ANC, y=value, group=name, fill=name)) + ylim(0, yscale) +
    geom_bar(stat = "identity", position = position_dodge()) + 
    geom_text(aes(label=value), position=position_dodge(0.8), size=3, vjust=-0.35) +
    scale_fill_manual(values=c(palettes$NatComms[1:2]),name="Match") +
    ylab("Frequency") + xlab("Genotype") +
    theme(axis.text.x = element_text(color = "black", angle = 35)) +
    theme(axis.text.y = element_text(color = "black"),
          legend.text = element_text(color = "black"),
          legend.title = element_text(color = "black"),
          legend.position = "bottom") +
    ggtitle(paste(g))
}

# genotype comparison heatmap
plot_geno_compare <- function(g) { 
  data.frame(
    as.data.frame(prop.table(table(df %>% select(starts_with(g))), 1)),
    N=as.data.frame(table(df %>% select(starts_with(g))))$Freq) %>%
    rename_with(~gsub(".*ALL", "ALL", gsub(paste0(".*",ANC), "ANC", .))) %>% 
    ggplot(aes(x=ANC, y=ALL)) + 
    geom_tile(aes(fill=Freq), color="black") + ylab("ALL") + xlab(ANC) +
    scale_fill_gradient2(limits=c(0,1), midpoint = 0.5, breaks=seq(0,1,0.25),
                         low=palettes$HeatMapYlRd[1], mid=palettes$HeatMapYlRd[35], 
                         high=palettes$HeatMapYlRd[100], name="Prop") + 
    geom_text(aes(label=N), size=3) + 
    theme(axis.text.x = element_text(color="black", angle=35, vjust=0.75),
          axis.text.y = element_text(color="black"),
          legend.position = "bottom") +
  ggtitle(g)
}

```

\n
\n

##### Genotype matching between ancestry
```{r plot-geno-match, fig.width=12, fig.height=9}
ggarrange(plot_geno_match(genotypes[1]), plot_geno_match(genotypes[2]), plot_geno_match(genotypes[3]), 
          plot_geno_match(genotypes[4]), plot_geno_match(genotypes[5]), plot_geno_match(genotypes[6]),
          plot_geno_match(genotypes[7]), plot_geno_match(genotypes[8]), plot_geno_match(genotypes[9]),
          nrow = 3, ncol = 3, common.legend = T)

ggarrange(
  ggarrange(plot_geno_match(genotypes[10]), plot_geno_match(genotypes[11]), 
            ncol=2, nrow=1, common.legend = T),
  ggarrange(plot_geno_match(genotypes[12]), plot_geno_match(genotypes[13]), 
            ncol=2, nrow=1, widths = c(1.5, 1), common.legend=T),
  nrow=2, common.legend = F)
```

#### Genotype comparison between ancestry 
```{r plot_geno-compare, fig.width=12, fig.height=8}
ggarrange(
  ggarrange(plot_geno_compare(genotypes[10]), plot_geno_compare(genotypes[11]), ncol=2, nrow=1, common.legend = T),
  ggarrange(plot_geno_compare(genotypes[12]), plot_geno_compare(genotypes[13]), ncol=2, nrow=1, widths = c(1.5, 1), common.legend=T),
  nrow=2, common.legend = F)
```

\n

### Compare associations of genotypes with salt intake frequency

```{r compare-salt-fn-1, fig.height=4, fig.width=12}
# By hand, set reference groups for lm to CAT and CAT/CAT (nontaster)
df.ref <- df %>%
  mutate(across(starts_with("haplo"), ~factor(.)),
         across(starts_with("diplo"), ~factor(.)),
         across(starts_with("taster"), ~factor(.))) %>%
  mutate(across(starts_with("haplo"), ~relevel(., ref="CAT")),
         across(starts_with("diplo"), ~relevel(., ref="CAT/CAT")),
         across(starts_with("taster"), ~relevel(., ref="Nontaster")))

# compare salt ~ geno
g_salt_lm_comp <- function(g)  {
  do.call(rbind, lapply(ANCs, function(ANC) {
    dat <- df.ref %>% select(starts_with(g), starts_with("addsalt_freq_QT"), starts_with(gPCs)) %>%
      mutate(across(starts_with(g), ~as.factor(.)))
    m<-lm(formula(paste0("addsalt_freq_QT.", ANC, "~", g, ".", ANC, "+", 
                           paste0(paste0(gPCs, ".", ANC), collapse = "+"))), data=dat)
    res<-rbind.data.frame(c(0,0), summary(m)$coef[startsWith(names(m$coef),g),c(1,2)]) ; rownames(res) <- NULL
    res$F_stat<- rep(anova(m)$`F`[1], length(m$xlevels))
    res<-cbind.data.frame(anc=rep(ANC,length(m$xlevels)), geno=rep(g,length(m$xlevels)), lvl=rep(m$xlevels[[1]]), res)
    colnames(res)[4:5]<-c("beta", "se")
    res %>% mutate(across(c(beta, se), ~as.numeric(.))) } ))
} ; all_g_salt_lm_comp <- do.call(rbind.data.frame, lapply(genotypes, g_salt_lm_comp))

# function to plot salt ~ geno
plot_lm_g_salt <- function(g) {
  yscale=max((all_g_salt_lm_comp %>% filter(geno==g) %>% mutate(y=abs(beta)+1.96*se))$y)*1.10
  all_g_salt_lm_comp %>%
    filter(geno == g) %>%
    ggplot(aes(x=lvl, y=beta, ymin=beta-1.96*se, ymax=beta+1.96*se, group=anc, color=anc)) + 
    geom_hline(yintercept = 0, color = "black") +
    geom_point(position=position_dodge(0.6), size=3) + 
    geom_errorbar(position=position_dodge(0.6), width=0.2) + 
    ylab("Beta (95% CI)") + xlab("Genotype") + ylim(-max(yscale), max(yscale)) +
    scale_color_manual(values=palettes$NatComms[3:4], name="Ancestry") + 
    theme(axis.text.x = element_text(color = "black", angle = 35),
          axis.text.y = element_text(color = "black"),
          legend.position = "bottom") + 
    ggtitle(g) 
}
```

###### F-statistic comparison
```{r compare-salt-F, out.width=6, out.height=3}
unique(all_g_salt_lm_comp %>% group_by(anc) %>%
  select(-c(lvl, beta, se)) ) %>% 
  pivot_wider(names_from="anc", values_from="F_stat") %>%
  kable(cation="F-test for taster status & salt frequency by ancestry") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F, position = "left")
```

```{r plot-salt-lm, fig.width=9, fig.height=9}
ggarrange(plot_lm_g_salt(genotypes[1]), plot_lm_g_salt(genotypes[2]), plot_lm_g_salt(genotypes[3]), 
          plot_lm_g_salt(genotypes[4]), plot_lm_g_salt(genotypes[5]), plot_lm_g_salt(genotypes[6]), 
          plot_lm_g_salt(genotypes[7]), plot_lm_g_salt(genotypes[8]), plot_lm_g_salt(genotypes[9]),
          nrow=3, ncol=3)
```


```{r summarise-geno, include=F} 
## HOLD ##
df.anc <- df %>% select(id, starts_with(genotypes)) %>%
  mutate(across(starts_with(genotypes), ~ as.character(.))) %>%
  pivot_longer(-id, names_to = "genotype.anc") %>% separate_wider_delim(genotype.anc, ".", names=c("genotype", "anc")) %>%
  pivot_wider(names_from=genotype) %>%
  group_by(anc)

df.anc %>% 
  reframe(
    #rs713
    rs713598_G_0=n_pct(rs713598_G, level="0"),
    rs713598_G_1=n_pct(rs713598_G, level="1"),
    rs713598_G_2=n_pct(rs713598_G, level="2"),
    #rs172
    rs1726866_G_0=n_pct(rs1726866_G, level="0"),
    rs1726866_G_1=n_pct(rs1726866_G, level="1"),
    rs1726866_G_2=n_pct(rs1726866_G, level="2"),
    #rs172
    rs10246939_C_0=n_pct(rs10246939_C, level="0"),
    rs10246939_C_1=n_pct(rs10246939_C, level="1"),
    rs10246939_C_2=n_pct(rs10246939_C, level="2"),
    #haplos
    haplo_0_CAT=n_pct(haplo_0, level="CAT"),
    haplo_0_GGC=n_pct(haplo_0, level="GGC"),
    haplo_0_CAC=n_pct(haplo_0, level="CAC"),
    haplo_0_CGC=n_pct(haplo_0, level="CGC"),
    haplo_0_GAC=n_pct(haplo_0, level="GAC"),
    haplo_0_GAT=n_pct(haplo_0, level="GAT"),
    haplo_1_CAT=n_pct(haplo_0, level="CAT"),
    haplo_1_GGC=n_pct(haplo_0, level="GGC"),
    haplo_1_CAC=n_pct(haplo_0, level="CAC"),
    haplo_1_CGC=n_pct(haplo_0, level="CGC"),
    haplo_1_GAC=n_pct(haplo_0, level="GAC"),
    haplo_1_GAT=n_pct(haplo_0, level="GAT"),
    #taster-status
    Taster_Status_Nontaster=n_pct(taster_status, level="Nontaster"),
    Taster_Status_Taster=n_pct(taster_status, level="Taster"),
    Taster_Status_Supertaster=n_pct(taster_status, level="Supertaster")) %>%
  t() %>% kable()
```


```{r compare-salt-fn-2, include=F}
# continuous genotypes
g_dose_salt_anc <- do.call(rbind.data.frame, lapply(ANCs, function(ANC) {
  do.call(rbind.data.frame, lapply(genotypes[1:9], function(g) {
    m<-summary(lm(formula(paste0("addsalt_freq_QT.", ANC, "~", g, ".", ANC, "+", 
                                 paste0(paste0(gPCs, ".", ANC), collapse = "+"))), data=df) )
    cbind(anc=ANC, geno=g, est=m$coef[2,1], p=m$coef[2,4], r2=m$r.squared, adjr2=m$adj.r.squared) })) 
}))

# categorical genotypes
g_cat_salt_anc <- do.call(rbind.data.frame, lapply(ANCs, function(ANC) {
  do.call(rbind.data.frame, lapply(genotypes[10:13], function(g) {
    m<-lm(formula(paste0("addsalt_freq_QT.", ANC, "~", g, ".", ANC, "+", 
                         paste0(paste0(gPCs, ".", ANC), collapse = "+"))), data=df)
    cbind(anc=ANC, geno=g, est=anova(m)$`F value`[1], p=anova(m)$`Pr(>F)`[1], 
          r2=summary(m)$r.squared, adjr2=summary(m)$adj.r.squared) } )) })) 

# combine
g_salt_anc <- rbind.data.frame(g_dose_salt_anc, g_cat_salt_anc) %>%
   mutate(adjr2=as.numeric(adjr2)) %>% select(anc, geno, adjr2) %>%
    mutate(geno=factor(geno, levels=c(genotypes))) %>%
    mutate(geno_group = gsub("_.*", "", geno)) %>%
    mutate(geno_group = factor(geno_group, levels=c("rs713598", "rs1726866", "rs10246939", "haplo", "diplo", "taster")))

# function to plot barplots
plot_r2_g_salt <- function(g) {
  yscale <- max((g_salt_anc %>% filter(geno==g))$adjr2)*1.10
  g_salt_anc %>% filter(geno == g) %>% 
  ggplot(aes(x=geno, y=adjr2-0.004, group=anc, fill=anc)) +
  geom_bar(position = position_dodge(), stat = "identity") +
  scale_fill_manual(values=palettes$NatComms[3:4], name="Ancestry") + 
  ylab("Model Adjusted R2") + xlab("Genotype") + #ylim(0, yscale) +
  theme(axis.text.x = element_text(color = "black", size=8), 
        axis.text.y = element_text(color = "black", size=8),
        legend.position = "bottom") +
  ggtitle(g) 
}
```

##EOF

