---
title: "Manuscript"
output: html_document
date: "2025-04-07"
---


```{r setup, include=FALSE, echo=F, error=F, warning=F, message=F}
knitr::opts_chunk$set(echo = F, error = F, warning = F, fig.align="center", 
                      out.width = "90%", out.height = "50%", fig.height=5)

#BiocManager::install("ComplexHeatmap")
library("ComplexHeatmap")

lapply(c("tidyverse", "data.table", "paletteer", "ggpubr", "ggtext", #"gplots", "R3port", 
         "xtable", "knitr", "tinytex", "emmeans"),  library, character.only = TRUE)

lapply(list.files("../../pantry/functions/", full.names = T), source)
source("../scripts/pantry.R", echo=F)
```

```{r}
#Diabetologia Color Palette: https://diabetologia-journal.org/wp-content/uploads/2018/07/Colour-palette-for-flow-charts-and-graphs.pdf
rgb2hex <- function(r,g,b) {
  RGB=c(r,g,b) ; maxRGB=max(RGB)
  hex=rgb(r,g,b,maxColorValue = 255)
  return(hex)
}

diabetologia_palette <- c(rgb2hex(242,222,196), rgb2hex(138,169,214), rgb2hex(76,89,136), rgb2hex(181,180,180))
nature_palette <- palettes$nature_diplos_dominant[c(1,3:5)]
haplo_palette <- nature_palette[c(1,2,4)] #taste_palette[c(1,2,4)]
diplo_palette <- nature_palette#taste_palette
snpdose_palette <- c(rgb2hex(170,196,226), rgb2hex(120,131,186), rgb2hex(96,108,158))

study_palette <- c(UKBB=as.vector(diplo_palette)[3], MAGIC=rgb2hex(192,65,64))


#figure theme
snp_theme <- theme_bw() + 
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        axis.text = element_text(color="black"),
        legend.key.size = unit(0.5, "cm"),
        legend.spacing.y = unit(0.1,"cm"),
        legend.margin = margin(0.5,0.5,0.5,0.5),
        legend.key.spacing = unit(0.1, "cm"),
        legend.title = element_text(face="bold"),
        legend.position = "right")


#Example markup figures: https://diabetologia-journal.org/wp-content/uploads/2023/08/Fig-mark-up-examples.pdf
```

```{r}
###############################
##   Load data by ancestry   ##
###############################

# system arguments
EUR=c("European"="EUR")
tag="v6"

# load postprocessed dataframe
postprocessed <- readRDS("../data/processed/ukb_postprocessed_EUR_20241227.rda")

analysis <- readRDS(paste0("../data/processed/ukb_analysis_EUR.rda"))
range(analysis$glu) #: 2.116 - 8.002
  
#load dietPCs
dietPCs.dat <- readRDS(paste0("../data/processed/diet/ukb_EUR_dietPCs.rda"))
dietPC_emm <- read.csv(paste0("../data/processed/descr_tab_emm_dietPC_diplos_sexageac_", tag,".csv"))
dietPC_pstat <- read.csv(paste0("../data/processed/descr_tab_teststat_dietPC_diplos_sexageac_",tag,".csv"))

taste_diplos.labs <- c("AVI/AVI", "AVI/PAV", "PAV/PAV")
taste_diplos <- c("AVI/AVI", "AVI/PAV", "PAV/PAV")
taste_diplos_005.labs <- c(taste_diplos.labs, "AVI/AAV", "AAV/PAV")

## Add glucose values with mg/dl
analysis <- analysis %>% mutate(
  glu_mgdl = glu*18.018,
  glu2hr_mgdl = glu2hr*18.018)

```

```{r}
## Housekeeping

# Bitter SNPs
bitter_snps <- c("rs713598_G", "rs1726866_G", "rs10246939_C", "rs10772420_A", "rs2597979_G")

# Build Models
m1_base = "age+sex+gPC1+gPC2+gPC3+gPC4+gPC5+gPC6+gPC7+gPC8+gPC9+gPC10+ac.f+fast_cat"
m2_bmi = paste0(m1_base, "+bmi")
m3_lifestyle = paste0(m2_bmi, "+smoke_level.lab+physact_level.lab+alch_freq.lab")
m4_diet = paste0(m3_lifestyle, "+dietPC3+dietPC4+dietPC6+dietPC11+dietPC12+dietPC13+dietPC15+dietPC16+dietPC17")
models.l = list("Base"= m1_base, "BMI"=m2_bmi, "Lifestyle"=m3_lifestyle, "Diet.Patterns"=m4_diet)
models.nofast.l <- as.list(gsub("[+]fast_cat", "", models.l))
names(models.nofast.l) <- names(models.l)

diet_vars.labs <- c(
  raw_veg_QT="Raw vegetable intake",
  cooked_veg_QT="Cooked vegetable intake",
  fresh_fruit_QT="Fresh fruit intake",
  dried_fruit_QT="Dried fruit intake",
  oily_fish_QT="Oily fish intake",
  nonoily_fish_QT="Non-oily fish intake",
  procmeat_QT="Processed meat intake",
  poultry_QT="Poultry intake",
  cheese_QT="Cheese intake",
  beef_QT="Beef intake",
  lamb_QT="Lamb intake",
  pork_QT="Pork intake",
  bread_type_white_vs_brown_or_whole_BIN="Prefer white>brown/whole bread",
  bread_intake_QT="Bread intake",
  milk_type_full_vs_low_or_nonfat_BIN="Prefer full>low/no-fat milk",
  cereal_type_sugar_vs_any_bran_BIN="Prefer sugary>bran cereal",
  coffee_type_decaf_vs_regular_BIN="Prefer decaf>caffeinated coffee",
  cereal_intake_QT="Cereal intake",
  spread_type_butter_vs_any_other_BIN="Prefer butter>other spreads",
  coffee_QT="Coffee intake",
  tea_QT="Tea intake",
  water_QT="Water intake",
  addsalt_always_often_vs_nrs_BIN="Always/often add salt",
  hotdrink_temp_hot_or_vhot_vs_warm_BIN="Prefer hot>warm drink temps",
  smoker_current_vs_never_BIN = "Current (vs never) smoker",
  smoker_former_vs_never_BIN = "Former (vs never) smoker",
  physact_low_vs_high_BIN = "Low (vs high) PA level",
  physact_mod_vs_high_BIN = "Moderate (vs high) PA level",
  alch_daily_rarelynever_BIN = "Daily (vs rarely/never) drink alcohol",
  alch_weekly_rarelynever_BIN = "Weekly (vs rarely/never) drink alcohol",
  alch_monthly_rarelynever_BIN = "Monthly (vs rarely/never) drink alcohol"
)

```

\n


## Figure 1. Distribution of TAS2R38 Haplotypes & Diplotypes 
```{r, error=F, message=F, fig.height==4, fig.width=6}
# ================================
## Summary of Haplotypes
# ================================

## haplo_legend
pal_haplo_legend <- haplo_palette
names(pal_haplo_legend) <- paste(c("Nontaster", "Taster", "Non-canonical"), "haplotype")

tab_haplos_AA <- analysis %>%
  select(id, haplo_0_aa, haplo_1_aa) %>%
  pivot_longer(-id, values_to="haplo") %>%
  group_by(haplo) %>% reframe(n=n(), pct=(n()/nrow(.))*100) %>%
  arrange(n) %>% mutate(haplo_order = factor(haplo, levels=haplo[order(n, decreasing = T)])) %>%
  mutate(diplo_legend = factor(ifelse(haplo_order == "AVI", "Nontaster haplotype", 
                                      ifelse(haplo_order == "PAV", "Taster haplotype", "Non-canonical haplotype")), 
                               levels=paste(c("Nontaster", "Taster", "Non-canonical"), "haplotype"))) %>%
  #Filter to "common" haplotypes (>1% of the cohort; >0.005, frequency)
  filter(pct>0.05) ; plot_haplos_AA <- tab_haplos_AA %>% 
  ggplot(aes(x=haplo_order, y=pct, fill=diplo_legend)) + 
  geom_bar(stat = "identity") + 
  scale_x_discrete(name = "**_TAS2R38_ Haplotype**") + 
  scale_y_continuous(limits=c(0,62), breaks=seq(0,60,10)) +
  ylab("Frequency (%)") + 
  scale_fill_manual(values = pal_haplo_legend, name=expression(italic("TAS2R38") ~ "Haplotypes (n)")) + 
  ggtheme_fancy + 
  theme(axis.text.x = element_text(size=8, color = "black", face="italic"), #angle=25, hjust=1, 
        axis.title.x = element_markdown(),
        axis.title = element_text(color = "black", face = "bold", size=10),
        axis.text.y = element_text(size=8, color = "black"),
        
        legend.position = "inside",
        legend.text = element_text(size=8), legend.title=element_blank(),
        legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
        legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
        legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
          
        legend.key.size = unit(1, 'line'),
        legend.margin = margin(3,3,3,3, unit="pt")
        ) +
  
    geom_text(aes(label=n), vjust=-0.35, size=3.0)

plot_haplos_AA

# ================================
## Summary of Diplotypes
# ================================

## Create variable in analysis for **common diplotypes**
diplos_gt005_AA <- names(which(prop.table(table(analysis$diplos_common_AA))>0.005))
analysis <- analysis %>% mutate(
  diplos_AA = factor(ifelse(diplos_common_AA %in% diplos_gt005_AA, diplos_common_AA, NA), levels=c(diplos_gt005_AA)))

## diplo_legend
pal_diplo_legend <- paste(c("Nontaster", "Taster", "Supertaster", "Non-canonical"), "diplotype")
names(pal_diplo_legend) <- paste(c("Nontaster", "Taster", "Supertaster", "Non-canonical"), "diplotype")

## Table
tab_diplos_AA <- analysis %>% 
  filter(complete.cases(diplos_AA)) %>% #diplos_common_AA %in% diplos_common_AA) %>% 
  reframe(n=table(diplos_AA), pct=prop.table(table(diplos_AA)), diplos_AA=names(table(diplos_AA))) %>%
  arrange(-n) %>% 
  mutate(color = factor(ifelse(diplos_AA == "AVI/AVI", "Nontaster",
                                 ifelse(diplos_AA == "AVI/PAV", "Taster",
                                        ifelse(diplos_AA == "PAV/PAV", "Supertaster", "Non-canonical"))),
                          levels=c("Nontaster","Taster","Supertaster","Non-canonical"))) %>%
  mutate(diplos_AA_order=factor(diplos_AA, levels=c("AVI/AVI", "AVI/PAV", "PAV/PAV", "AVI/AAV", "AAV/PAV")),
         diplo_legend = factor(paste(color, "diplotype"), levels=
                                 paste(c("Nontaster","Taster","Supertaster","Non-canonical"), "diplotype"))) %>%
  filter(pct>0.005) ; plot_diplo_AA <- tab_diplos_AA %>%
  ggplot(aes(x=diplos_AA_order, y=pct*100, fill=diplo_legend)) + 
  geom_bar(stat = "identity") + scale_x_discrete(name = "**_TAS2R38_ Diplotypes**") + 
  scale_y_continuous(limits=c(0,52), breaks=seq(0,50,10)) + ylab("Frequency (%)") + 
  scale_fill_manual(values = pal_diplo_legend, name=expression(italic("TAS2R38") ~ "Diplotypes (n)")) + 
  ggtheme_fancy +
    theme(axis.text.x = element_text(size=8, color = "black", face="italic"), #angle=25, hjust=1, 
          axis.text.y = element_text(size=8, color = "black"),
          axis.title.x = element_markdown(),
          axis.title = element_text(color = "black", face = "bold", size=10),
          
          legend.position = "inside",
          legend.text = element_text(size=8), legend.title=element_blank(),
          legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
          legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
          legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
          
          legend.key.size = unit(1, 'line'),
          legend.margin = margin(3,3,3,3, unit="pt"),
          
          ) +  # Optional: Add background
  
  geom_text(aes(label=n), vjust=-0.35, size=3.0) ; plot_diplo_AA


# =================================================
# Panel plot of TAS2R38 haplotypes & diplotypes 
# =================================================
plot_haplo_diplo_AA <- ggarrange(plot_haplos_AA, "",  plot_diplo_AA, ncol=3, 
                                 widths=c(0.75, 0.05, 1), align = c("hv"), labels=c("A", "", "B")
                                 )
plot_haplo_diplo_AA
plot_haplo_diplo_AA #%>% ggsave(filename=paste0("../output/manuscript/descr_plot_genoAA005_panel_",tag,"_v2.pdf"), 
                     #          height=3.75, width=9.15)
 
```


## Table 1. Descriptive characteristics 
```{r, include=T}
analysis <- analysis %>% filter(!is.na(taste_diplos))
read.csv(paste0("../data/processed/descr_tab_diplo_AA_taste_",tag,".csv")) %>% kable(caption="Descriptive Table 1")

## Tabulate table 1 with median (IQR)
analysis %>% group_by(taste_diplos) %>%
   filter(complete.cases(taste_diplos)) %>%
  reframe("Age, years"=median_IQR(age, d=1),
          "BMI, kg/m2"=median_IQR(bmi, d=1),
          "Cigarettes per day"=median_IQR(cigarettes_per_day, d=1),
          "PA level"=median_IQR(physact_met_excess, d=1),
          "Drinks per week"=median_IQR(alch_drinks_per_week, d=1),
          "Raw vegetables, tbsp/week"=median_IQR(raw_veg_QT, d=2),
          "Coffee, cups/week"=median_IQR(coffee_QT, d=2),
          "Tea, cups/week"=median_IQR(tea_QT, d=2))

## re-calculate means with cigarette smokers, excluding outliers >100
summary(lm(cigarettes_per_day~taste_diplos.num,data=analysis %>% filter(cigarettes_per_day<60)))

## P-values for UNADJUSTED Chi.Square or P-trend tests (with continuous ) for Table 1 *********
table1_pvals <- do.call(rbind.data.frame, lapply(1:length(vars_to_summarize), function(v) {
  dat <- analysis %>% select(taste_diplos.num, var=names(vars_to_summarize[v]))
  if(is.numeric(dat$var) ==T) {
    cbind.data.frame(Var=vars_to_summarize[[v]], Pvalue=summary(lm(var~taste_diplos.num, data=dat))$coef[2,4])
  } else {
    cbind.data.frame(Var=vars_to_summarize[[v]], Pvalue=chisq.test(dat$taste_diplos.num, dat$var)$p.value)
  }
})) %>% mutate(P_formatted = format_p.fun(Pvalue))
      
table1_pvals %>% write.csv("../output/manuscript/table1_pvalues.csv")
# NOTE: P-value for added salt is wrong...
```


##### Figure 2. Plot of estimated marginal means & replication 
```{r}

## PANEL A: EMM for glucose x fasting time by DIPLOTYPE  -------------------------
emm_gluXfast <- read.csv(paste0("../data/processed/res_tab_emm_gluXfast_6to12_tastediplo_mBMI_nofast_mgdl_",tag,".csv"))
emm_allglycemic <- read.csv(paste0("../data/processed/res_tab_emm_allglycemic_tastediplos_mgdl_", tag, ".csv"))

# NOTE: looking deeper into the 12-24 range showed a TON of variability, so the EMMs just focused on the 6-12 window
emm_glu_mgdl_bmi <- rbind.data.frame(
  emm_allglycemic %>% filter(outcome == "Glucose (mg/dL)" & model == "BMI") %>% 
    mutate(fast = "Random", .after="model"), emm_gluXfast)


## PANEL B: EMM for glucose x fasting time by EACH SNP -------------------------
emm_bittersnp_x_gluFast <- read.csv(paste0("../data/processed/sens_tab_emm_gluXfast_bittersnps_mBMI_mgdl_",tag,".csv"))
bitter_snps_cat <- paste0(bitter_snps, ".cat")
snps_fancy <- c("rs713598 (C>G)", "rs1726866 (G>A)", "rs10246939 (T>C)")
snps_fancy_2row <- c("rs713598\n(C>G)", "rs1726866\n(G>A)", "rs10246939\n(T>C)")

## PANEL C: Beta (95% CI) per SNP for REPLICATION -------------------------

#negative control data
snpukb <- read.csv(paste0("../data/processed/sens_tab_lm_gluXfast_bittersnps_mBMI_mgdl_",tag,".csv")) %>%
  filter(model  == "BMI" & fast == "0to2hr") %>%
  mutate(Bitter_SNP = factor(exposure, levels=rev(bitter_snps))) %>%
  filter(Bitter_SNP %in% c("rs713598_G", "rs1726866_G", "rs10246939_C")) %>%
  mutate(snp = c("rs713598 (C>G)", "rs1726866 (G>A)", "rs10246939 (T>C)")) %>%
  mutate(Study="UKBB (0-2hr)") %>%
  select(Study, n, snp, beta, se, p)

#independent replication data
snpmagic <- read.csv("../data/meta_analysis_2hg.csv") %>%
  filter(Trait == "0to2hrG_adjBMI") %>%
  select(snp=SNP, beta=BETA, se=SE, n=N, p=P_value) %>%
  mutate(across(c("beta", "se", "n"), ~as.numeric(.))) %>%
  mutate(Bitter_SNP = factor(c("rs713598_G", "rs1726866_G", "rs10246939_C")),
         Study = "MAGIC (2-hr\nOGTT)") %>%
  mutate(across(c("beta", "se"), ~.*18.018)) %>%
  select(Study, n, snp, beta, se, p)

snplvl <- rbind.data.frame(snpukb, snpmagic) %>%
  bind_rows(cbind.data.frame(Study="UKBB (0-2hr)", n=0, snp="TAS2R38 Diplotype", beta=-0.23, se=0.07653061,p=0.003)) %>%
  mutate(snp=factor(snp, levels=c(rev(snps_fancy), "TAS2R38 Diplotype"))) %>%
  mutate(Study=factor(Study,levels=c("MAGIC (2-hr\nOGTT)", "UKBB (0-2hr)"))) %>% 
  mutate(geno = c(rep("Variant",6), "Diplotype")) ; snplvl

# ===========
# panel A 
# ===========
panelA <- emm_glu_mgdl_bmi %>%
  filter(fast != "12+hr") %>%
  mutate(fast=factor(fast, levels=rev(c("6to12hr", "5hr", "4hr", "3hr", "0to2hr", "Random")),
                     labels=rev(c("6-12 hr", "5 hr", "4 hr", "3 hr", "0-2 hr", "Anytime from\n0-12 hrs"))),
         Diplotypes = factor(level, levels=c("AVI/AVI", "AVI/PAV", "PAV/PAV"))) %>%
  mutate(type = factor(ifelse(fast=="Anytime from\n0-12 hrs", "Random", "Fasting"), levels=c("Random", "Fasting"))) %>%
  ggplot(aes(y=emmean, x=fast, ymin=lowCI, ymax=upCI, color=Diplotypes)) + 
  #facet_grid(~type, scale="free", space="free_x", labeller = as_labeller(c("Random" = "Random Glucose", "Fasting"="Glucose by Self-Reported Fasting Time"))) +
  facet_grid(~type, scale="free", space="free_x", labeller = as_labeller(c("Random" = "", "Fasting"=""))) +
  scale_y_continuous(limits=c(87.25, 90), breaks=seq(87.5,90,0.5)) +
  geom_point(size=1.75, position=position_dodge(0.35)) + 
  geom_errorbar(width=0.15, position=position_dodge(0.35), linewidth=0.45) + 
  scale_color_manual(values=diplo_palette, name="**_TAS2R38_ Diplotype**") +
  ylab("Glucose level (mg/dL)\n") + xlab("Self-reported fasting window (hr)") + 
  theme_bw() +
  theme(#legend.position = "right",
        axis.title.x = element_text(size=8, face="bold", vjust=3), 
        axis.title.y = element_text(size=8, face="bold", vjust=-2), 
        axis.text.x = element_text(size=7, color="black"), 
        axis.text.y = element_text(size=7),
        panel.grid.major.y = element_line(linewidth = 0.25), 
        strip.background = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill="#00000000"),
        strip.text = element_text(face="bold", size=8), 
        
        legend.position = "inside",
        legend.title = element_markdown(size=8, margin = margin(0,0,2.5,0)), #element_text(face="bold", size=9),
        legend.text = element_text(size=7, face="italic"), #legend.title=element_blank(),
        legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
        legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
        legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
          
        legend.key.size = unit(0.65, 'line'),
        legend.margin = margin(2,2,2,2, unit="pt"),
        plot.margin=margin(1,4,1,1, unit="line")
  ) ; panelA

# ===========
# panel B 
# ===========
panelB <- emm_bittersnp_x_gluFast %>% 
  as.data.frame() %>%
  dplyr::mutate(fast = factor(fast, levels=c("6+hr", "5hr", "4hr", "3hr", "0to2hr"),
                       labels =c(">6 hrs", "5 hrs", "4 hrs", "3 hrs", "0-2 hrs")),
         Genotypes = factor(level, levels=c("0_0","0_1", "1_1"), labels=c("00", "01", "11")), .before=outcome) %>%
  filter(exposure %in% c("rs713598_G.cat", "rs1726866_G.cat", "rs10246939_C.cat")) %>%
  mutate(SNP = rep(bitter_snps[1:3], each=15), .before=level) %>%
  mutate(SNP=factor(SNP, levels=bitter_snps[1:3], labels=snps_fancy_2row)) %>%
  mutate(Gene = c(rep("TAS2R38-PROP/PTC", 15*3))) %>%
  filter(fast == "0-2 hrs") %>%
  ggplot(aes(y=emmean-88.5, x=SNP, ymin=lowCI-88.5, ymax=upCI-88.5, color=Genotypes)) + 
  scale_y_continuous(limits=c(0,2), breaks=seq(0,2,0.5), labels = sprintf("%.1f", seq(0,2,0.5)+88.5)) + #
  geom_errorbar(position=position_dodge(0.35), linewidth=0.35, width=0.15) + 
  geom_point(size=1.75, position=position_dodge(0.35), shape=17) + 
  scale_color_manual(values=c(as.vector(diplo_palette)), name=paste0("Genotype for\nBitter Allele")) + 
  ylab("Glucose level (mg/dL)") + xlab("**_TAS2R38_** functional variants") +
  ggtitle("0-2 hr Glucose, mg/dL ") +
  guides(color=guide_legend(override.aes = list(size=1.5))) +
  theme_bw() +
  theme(axis.title.y = element_text(size=7, face="bold", margin = margin(r=5)), 
        axis.title.x = element_markdown(size=7, face="bold", margin = margin(t=3)), 
        axis.text.x = element_text(size=6, color="black"), 
        axis.text.y = element_text(size=6, color="black"),
        panel.grid.major.y = element_line(linewidth = 0.25), 
        strip.background = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill="#00000000"),
        legend.position = "inside",
        legend.title = element_text(face="bold", size=6, margin = margin(0,3.5,0,0)), #element_text(face="bold", size=9),
        legend.text = element_text(size=6), #legend.title=element_blank(),
        legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
        legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
        legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
        legend.key.size = unit(0.5, 'line'),
        legend.spacing.y = unit(0.01,"cm"),
        legend.margin = margin(2,2,2,2, unit="pt"),
        legend.direction = "horizontal",
        plot.margin=margin(1,1,1.25,3,unit="line"),
        plot.title=element_text(size=8, hjust=0.5, margin=margin(b=1))
  ) ; panelB

# ===========
# panelC
# ===========
panelC <- snplvl %>%
  ggplot(aes(y=snp, x=beta, xmin=beta-1.96*se, xmax=beta+1.96*se, color=Study, shape=geno)) + 
  snp_theme + 
  geom_point(size=1.75, position=position_dodge(0.55)) + 
  geom_errorbarh(height=0.25, position=position_dodge(0.55), linewidth=0.35)  + 
  geom_vline(xintercept = 0) + 
  scale_x_continuous(limits = c(-0.75,0.25)) +
  scale_color_manual(values=as.vector(study_palette)) +
  scale_shape_manual(values=c(19,17), name="Genetic\nExposure") +
  ylab("") + xlab("Beta (95% CI)") +
  guides(color=guide_legend(reverse = T, override.aes = list(size=1)),
         shape=guide_legend(override.aes = list(size=1))) +
  ggtitle("Glucose level, mg/dL") +
  theme(
    legend.position = "right",
    legend.title = element_text(face="bold", size=5, margin = margin(0,0,2.5,0)), #element_text(face="bold", size=9),
    legend.text = element_text(size=5), #legend.title=element_blank(),
    #legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
    #legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
    legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
    legend.key.size = unit(0.5, 'line'),
    legend.margin = margin(1,2,2,1, unit="pt"),
    legend.spacing = unit(2,unit="line"),
    legend.box.margin = margin(0,0,0,-5),
    
    axis.title.x = element_text(size=7, face="bold", margin=margin(t=-3)), 
    axis.text.x = element_text(size=6), 
    axis.text.y = element_text(size=6),
    panel.grid.major.x = element_line(linewidth = 0.25), 
    panel.grid.major.y = element_blank(),
    strip.background = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.background = element_rect(fill="#00000000"),
    strip.text = element_text(face="bold", size=8),
    plot.margin=margin(1,1,1,1,unit="line"),
    plot.title=element_text(size=8, hjust=0.5, margin=margin(b=1))
  ) ; panelC

# ==================
## Combine panels
# ==================
panelBC <- ggarrange(panelB, panelC, nrow=1, labels=c("   B", " C"), align="h", font.label = list(size=12), vjust=3)
Fig2_allpanels <- ggarrange(panelA, panelBC, nrow=2, heights=c(2.75,2.25), labels = c(" A", " "), font.label = list(size=12), vjust=3)
Fig2_allpanels %>% ggsave(filename="../output/manuscript/figure2_allpanels_v2.pdf", height=5, width=7)

# for ADA summary report
Fig2_allpanels %>% ggsave(filename="../output/manuscript/figure2_allpanels_v2_thin.pdf", height=5, width=6.5)
```


### Figure 3. Plot of sensitivity analysis, adjusting for dietary and lifestyle habits

###### Supplementary Figure: Dot plot for TAS2R38 & waterfall for glucose significance 
```{r}
## BMI model
diplo_diet <- fread("../data/processed/sens_tab_lm_diplo_dietlife_mBMI_v6.csv")
panelA.supp <- diplo_diet %>%
  arrange(beta) %>%
   mutate(Color=ifelse(as.numeric(p) < .05 & beta<0, "A",
                      ifelse(p>0.05 & beta<0, "B", 
                             ifelse(p<0.05 & beta>0, "D", "C")))) %>%
  mutate(Diet=model) %>%
  mutate(Diet=factor(model, levels=Diet[order(beta)])) %>% 
  ggplot(aes(x=beta, xmin=beta-1.96*se, xmax=beta+1.96*se, y=Diet, color=Color)) + 
  geom_vline(xintercept = 0, linewidth=0.35) +
  geom_point(size=1) + 
  geom_errorbarh(height=0.35, linewidth=0.35) + 
  scale_color_manual(values=c(palettes$hm_palette_rwb[c(10,30,70,90)])) +
  #scale_x_continuous(limits=c(-2,2.5), breaks=seq(-2,2.5,1)) +
  theme_bw() + ylab("") +
  xlab("Beta (95% CI)") + 
  coord_flip() +
  ggtitle("TAS2R38 - Diet/Lifestyle Associations") +
  theme(panel.grid.minor = element_blank(),
       # panel.grid.major.y = element_line(linewidth=0.2),
       panel.grid.major.x = element_blank(),
       panel.grid.major.y = element_line(linewidth=0.2),
       legend.text = element_markdown(size=5),
       axis.text.y = element_text(color="black", size = 6),
       axis.text.x = element_text(color="black", size = 6, vjust=0.5, hjust=1, angle=90),
        axis.ticks = element_line(linewidth=0.1),
        axis.title = element_text(face="bold", size=6, margin=margin(t=3)),
        
        legend.position = "right",
        legend.title = element_text(face="bold", size=5, margin = margin(0,0,2.5,0)), #element_text(face="bold", size=9),
        legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
        legend.key.size = unit(0.5, 'line'),
        legend.margin = margin(1,1,2,1, unit="pt"),
        legend.spacing = unit(1,unit="line"),
        legend.box.margin = margin(0,0,0,-5),
       plot.title=element_text(size=8, margin=margin(b=1))
        ) ; panelA.supp

## waterfall: gluc ~ diet
panelB.supp <- diplo_2hrglu %>%
  mutate(Diet=model) %>%
  mutate(logp=-log10(p)) %>%
  mutate(Color=ifelse(beta<0 & p<.05/31, "A",
                      ifelse(p>0.05/32 & beta<0, "B", 
                             ifelse(beta>0 & p<.05/31, "D", "C")))) %>%
  mutate(dir=ifelse(beta<0,-logp,logp)) %>%
  mutate(Diet=factor(Diet, levels=Diet[order(dir)])) %>%
  ggplot(aes(x=logp, y=Diet, fill=Color)) + 
  geom_bar(stat="identity") + 
  geom_vline(xintercept = 0, color="black", linewidth=0.55) + 
  geom_vline(xintercept = -log10(0.05/24), linetype="dashed", color="black", linewidth=0.25) +
  geom_vline(xintercept = -log10(0.05), linetype="longdash", color="black", linewidth=0.25) +
  scale_fill_manual(values=c(palettes$hm_palette_rwb[c(10,30,70,90)])) +
  xlab("-log10(P)") + ylab("") +
  ggtitle("0-2 hr Glucose, mg/dL") +
  theme_bw() + 
  theme(panel.grid = element_blank(),
        axis.text.y=element_text(size=6, color="black"),
        axis.text.x=element_text(size=6, color="black"),
        axis.title = element_text(size=6, face="bold"),
        legend.position = "inside",
        legend.text = element_text(size=5), legend.title=element_blank(),
        legend.position.inside = c(0.99, 0.98),  # Moves legend to top-right inside plot
        legend.justification.inside = c(1, 1),  # Aligns legend's top-right corner
        legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
          
        legend.key.size = unit(1, 'line'),
        legend.margin = margin(3,3,3,3, unit="pt"),
        plot.margin=margin(1,1,2,1, unit="line"),
        plot.title=element_text(size=8, margin=margin(b=1))
      ) ; panelB.supp

ggarrange(panelA.supp, "", panelB.supp, widths=c(1.65, 0.05, 1.15),nrow=1, 
          labels = c("A", "", "B"), common.legend = T, legend = "bottom") %>% 
  ggsave(filename="../output/manuscript/plot_suppl_dietlife_covars.pdf", height=3.5, width=8)

```

## Figure 3

```{r}
fig3_theme <- theme(
  panel.grid.minor = element_blank(),
  panel.grid.major.y = element_blank(),
  axis.text.x=element_text(color="black",size=7),
  axis.text.y=element_text(color="black",size=7),
  axis.title = element_text(color="black", face="bold",size=7),
  plot.margin = margin(1,1,1,1, unit="line"),
  plot.title = element_text(size=7, margin=margin(b=0.1,unit="line"))
)

fig3_legend <- theme(
  legend.position = "inside",
  legend.text = element_markdown(size=6, margin=margin(r=5, unit="pt")), 
  legend.title=element_text(size=6, face="bold", margin=margin(b=2, unit="pt")),
  legend.position.inside = c(0.98, 0.02),  # Moves legend to bottom-right inside plot
  legend.justification = c("right", "bottom"),  # Aligns legend's top-right corner
  legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
  legend.key.size = unit(0.55, 'line'),
  legend.margin = margin(3,3,3,3, unit="pt")
)


# ===================================================================
## panel A: TAS2R38-Glucose associations with sequential adjustment
# ===================================================================

adj_life <- c("alch_freq.lab", "smoke_level.lab", "physact_level.lab")
adj_diet <- c(names(diet_vars.labs)[1:24])
adj_dietlife <- c(adj_life, adj_diet)

model_life <- paste0(models.nofast.l$BMI, "+", paste0(adj_life, collapse = "+"))
model_diet <- paste0(models.nofast.l$BMI, "+", paste0(adj_diet, collapse = "+"))
model_lifediet <- paste0(model_life, "+", paste0(adj_diet, collapse = "+"))

adj_models <- list("BMI Model\n(Primary)"=models.nofast.l$BMI, "Lifestyle\nAdjusted"=model_life, 
                   "Diet\nAdjusted"=model_diet, "Lifestyle+Diet\nAdjusted"=model_lifediet)

## Just check LRT for TAS2R38...
mRed <- lm(formula(paste0("glu2hr_mgdl~", adj_models$`Lifestyle+Diet`)), data=analysis %>% filter(complete.cases(taste_diplos)))
mFull <- lm(formula(paste0("glu2hr_mgdl~taste_diplos+", adj_models$`Lifestyle+Diet` )), data=analysis)
lrt <- lmtest::lrtest(mRed, mFull) ; cbind.data.frame(Model=names(adj_models)[4], pLRT = lrt$`Pr(>Chisq)`[2]) #pLRT = 0.0150


r2_diet_diplo <- do.call(rbind.data.frame, lapply(1:length(adj_models), function(m) {
  mBase <- summary(lm(formula(paste0("glu2hr_mgdl~", adj_models[[m]])),data=analysis))
  mTaste <- summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", adj_models[[m]] )), data=analysis))
  cbind.data.frame(Model=names(adj_models)[m], 
                   baseR2=mBase$adj.r.squared, tasteR2=mTaste$adj.r.squared,
                   beta=mTaste$coef[2,1], se=mTaste$coef[2,2], p=mTaste$coef[2,4]
                   )
})) ; r2_diet_diplo

#library(lmtest)
lrt_diet_diplo <- do.call(rbind.data.frame, lapply(2:length(adj_models), function(m) {
  complete <- analysis %>% select("glu2hr_mgdl", all_of(strsplit(adj_models[[m]], split="[+]")[[1]])) %>% 
    filter(complete.cases(.))
  mRed <- lm(formula(paste0("glu2hr_mgdl~", adj_models[[1]])), data=complete)
  mFull <- lm(formula(paste0("glu2hr_mgdl~", adj_models[[m]] )), data=complete)
  lrt <- lmtest::lrtest(mRed, mFull) ; cbind.data.frame(Model=names(adj_models)[m], pLRT = lrt$`Pr(>Chisq)`[2])
})) ; lrt_diet_diplo


## PLOTS 
panelA <- r2_diet_diplo %>% 
  mutate(diffR2 = (baseR2-r2_diet_diplo$baseR2[1])*100) %>%
  mutate(LRT = c(0.1, -log10(lrt_diet_diplo$pLRT))) %>%
  select(Model, LRT) %>%
  mutate(Model=factor(Model, levels=rev(c(names(adj_models))))) %>%
  ggplot(aes(y=Model, x=LRT)) + 
  geom_bar(stat="identity", width=0.6,fill= "grey") + 
  xlab("-log10(P) for LRT (vs BMI Model)") + ylab("") +
  theme_bw() + 
  geom_vline(xintercept = 0, linewidth=0.35) +
  geom_vline(xintercept = -log10(.05), linetype="longdash", linewidth=0.55) +
  ggtitle("Covariate Associations w/ 0-2hr Glucose") +
  fig3_theme + theme(
    plot.title = element_text(size=6.5, margin=margin(b=0.1,unit="line"))) ;  panelA

panelB <- r2_diet_diplo %>% 
  mutate(lowCI=beta-1.96*se, upCI=beta+1.96*se) %>%
  mutate(Model=factor(Model, levels=rev(c(names(adj_models))))) %>%
  ggplot(aes(y=Model, x=beta, xmin=lowCI, xmax=upCI)) + 
  geom_errorbarh(height=0.25, linewidth=0.55, color=diplo_palette[2]) + 
  geom_point(size=2, color=diplo_palette[2]) + 
  geom_vline(xintercept = 0) + 
  xlab("Beta (95% CI) for _TAS2R38_ Diplotype") + ylab("") + #ylab("\nCovariate Adjustments") +
  theme_bw() + 
  scale_x_continuous(limits=c(-0.5,0.1)) +
  ggtitle("_TAS2R38_ Associations with 0-2hr Glucose") +
  fig3_theme + theme(
    axis.title.x = element_markdown(size=7, color="black"),
    plot.title = element_markdown(size=6.5, margin=margin(b=0.1,unit="line"))); panelB

ggarrange(panelA, panelB, ncol=1, align="v") 
#panelA <- ggarrange(pC, pD, widths=c(0.75,0.5), common.legend = T, legend = "none", align="h") ; panelA
#panelA %>% ggsave(filename="../output/manuscript/plot_figure3_panelA.pdf", height=3, width=4)


# ======================================================================
## panel B: Negative Control variants associate similarly with diet
## traits that are associated with 0-2 hr Glucose
# ======================================================================

snp_labels <- c(rep("_TAS2R38_ Diplotype",3), "_TAS2R14_ Negative Control", "_TAS2R19_ Negative Control")
names(snp_labels) <- bitter_snps

bitter_foods <- c("tea_QT", "coffee_QT", "addsalt_3lvl.lab", "alch_freq.lab", "cereal_intake_QT", 
                  "spread_type_butter_vs_any_other_BIN", "bread_type_white_vs_brown_or_whole_BIN", 
                  "fresh_fruit_QT", "procmeat_QT")

## P-values for UNADJUSTED Chi.Square or P-trend tests (with continuous ) for Table 1 *********
negcontrol_pvals <- do.call(rbind.data.frame, lapply(bitter_snps, function(snp) {
  do.call(rbind.data.frame, lapply(1:length(bitter_foods), function(v) {
    dat <- analysis %>% select(SNP=snp, var=bitter_foods[v])
    if(is.numeric(dat$var) ==T) {
      cbind.data.frame(Var=bitter_foods[[v]], Pvalue=summary(lm(var~SNP, data=dat))$coef[2,4])
      } else {
        cbind.data.frame(Var=bitter_foods[[v]], Pvalue=chisq.test(dat$SNP, dat$var)$p.value)
        }
    })) %>% mutate(P_formatted = format_p.fun(Pvalue)) %>% mutate(SNP=snp)
}))

negcontrl_palette <- c(rs10772420_A=rgb2hex(109,165,103), rs2597979_G=rgb2hex(193,121,99),
                       rs713598_G=diplo_palette[[2]], rs1726866_G=diplo_palette[[2]], rs10246939_C=diplo_palette[[2]])

negcontrl_palette <- c(diplo_palette[[2]],rgb2hex(109,165,103), rgb2hex(193,121,99))
names(negcontrl_palette) <- snp_labels[c(1,4:5)]

panelC <- negcontrol_pvals %>%
  mutate(snp=factor(SNP, levels=rev(bitter_snps))) %>%
  mutate(diet=diet_vars.labs[Var]) %>% mutate(diet=ifelse(Var=="addsalt_3lvl.lab", "Frequency of adding\nsalt to food",
                                                          ifelse(Var=="alch_freq.lab", "Alcohol use frequency", diet))) %>%
  mutate(snp_label = factor(snp, levels=bitter_snps, labels=snp_labels)) %>%
  filter(diet != "Frequency of adding\nsalt to food") %>%
  mutate(diet = factor(diet, levels=rev(unique(diet[order(Pvalue)])))) %>%  
  ggplot(aes(x=-log10(Pvalue), y=diet, fill=snp_label, group=snp)) + 
  geom_bar(stat="identity", position=position_dodge(0.65), width=0.6) + 
  xlab("-log10(P)") + ylab("") + 
  theme_bw() + 
  scale_fill_manual(values=c(negcontrl_palette), name="Bitter Taste Variant") +
  geom_vline(xintercept = 0, linewidth=0.35) +
  geom_vline(xintercept = -log10(0.05), linetype="longdash", linewidth=0.25) +
  guides(fill=guide_legend(reverse = T)) +
  ggtitle("Diet/Lifestyle Traits Associated with 0-2hr Glucose") +
  fig3_theme + fig3_legend + 
  theme(axis.text.y=element_text(color="black", size=7)) ; panelC

#panelC %>% ggsave(filename="../output/manuscript/plot_figure3_panelC_v2.pdf", height=2.5, width=5)


# ======================================================================
## panel C: Associations of TAS2R38 and Negative Control variants with
## 0-2hr glucose, before and after adjusting for diet/lifestyle
# ======================================================================

## EMM in the fully-adjusted model
emm_bittersnps_glu2hr <- do.call(rbind.data.frame, lapply(bitter_snps_cat, function(snp) {
  do.call(rbind.data.frame, lapply(1:length(adj_models), function(m) {
    get_emm.fun(exposure = snp, outcome = "glu2hr_mgdl", covars = adj_models[[m]], 
                reference = "0_0", data=analysis)$emm })) %>%
    mutate(Model = names(adj_models)[m])
     })) ; emm_bittersnps_glu2hr <- emm_bittersnps_glu2hr %>% mutate(
  Model = rep(rep(names(adj_models), each=3),length(bitter_snps_cat))
)

emm_snps_glu <- emm_bittersnps_glu2hr %>% 
  mutate(model=Model) %>%
  filter(Model=="Lifestyle+Diet")

negcontl_dose_palette <- c(TAS2R38_0=diplo_palette[[1]], TAS2R38_1=diplo_palette[[2]], TAS2R38_2=diplo_palette[[3]],
                           TAS2R14_0=rgb2hex(152,201,139), TAS2R14_1=rgb2hex(109,165,103), TAS2R14_2=rgb2hex(99,122,98),
                           TAS2R9_0=rgb2hex(214,159,151), TAS2R9_1=rgb2hex(193,121,99), TAS2R9_2=rgb2hex(150,102,90))
legend_dose <- data.frame(
  Label = c("0 alleles", "1 allele", "2 alleles"),
  Color = c("#CCCCCC", "#7F7F7F", "#000000"),
  x = 1, y = 1
)

label = c("Line 1<br><i>Group A</i>", "Line 1<br><i>Group B</i>")

gene_labels <- c(paste0("<i>TAS2R38</i><br>(", bitter_snps[1:3], ")"), 
                 "<i>TAS2R14</i><br>(rs10772420_A)", "<i>TAS2R19</i><br>(rs2597979_G)")
names(gene_labels) <- bitter_snps 
gene_labels_lineheight <- paste0("<span style='line-height:1.5'>", gene_labels) ; names(gene_labels_lineheight)<-gene_labels

panelD <- emm_snps_glu %>%
  mutate(across(c(emmean, SE, lowCI, upCI), ~as.numeric(.))) %>%
  mutate(Genotypes = factor(level, levels=c("0_0","0_1", "1_1"), labels=c("00", "01", "11")), .before=outcome) %>%
  mutate(snp = factor(gsub(".cat", "", exposure), levels=bitter_snps)) %>%
  mutate(snp_label = case_when(snp %in% bitter_snps[1:3] & level=="0_0"~"TAS2R38_0",
                               snp %in% bitter_snps[1:3] & level=="0_1"~"TAS2R38_1",
                               snp %in% bitter_snps[1:3] & level=="1_1"~"TAS2R38_2",
                               snp == bitter_snps[4] & level=="0_0"~"TAS2R14_0",
                               snp == bitter_snps[4] & level=="0_1"~"TAS2R14_1",
                               snp == bitter_snps[4] & level=="1_1"~"TAS2R14_2",
                               snp == bitter_snps[5] & level=="0_0"~"TAS2R9_0",
                               snp == bitter_snps[5] & level=="0_1"~"TAS2R9_1",
                               snp == bitter_snps[5] & level=="1_1"~"TAS2R9_2")) %>%
  mutate(snp_label = factor(snp_label, levels=names(negcontl_dose_palette))) %>%
  mutate(gene_snp = factor(snp, levels=bitter_snps, labels=c(gene_labels))) %>%
  ggplot(aes(y=emmean-88.25, x=gene_snp, ymin=lowCI-88.25, ymax=upCI-88.25, color=snp_label)) + 
  scale_y_continuous(limits=c(-0.5,2), breaks=seq(-0.5,2,0.5), labels = seq(-0.5,2,0.5)+88) +
  geom_point(position=position_dodge(0.35), size=1.25, shape=17) + 
  geom_errorbar(position=position_dodge(0.35), linewidth=0.35, width=0.15) + 
  scale_color_manual(values=c(negcontl_dose_palette), name="Bitter Taste Variant", guide=NULL) + 
  ylab("Blood glucose (mg/dL)") + xlab("") +
  theme_bw() +
  fig3_theme + fig3_legend +
  theme(panel.grid.major.y = element_line(),
        panel.grid.major.x = element_blank(),
        axis.text.y = element_text(color="black", size=6),
        axis.text.x = element_markdown(color="black", size=6.5, lineheight=1.35),
        axis.title.y = element_text(color="black", size=7),
        #legend.position = "none",
        plot.margin=margin(0,1,1,2, unit="line"))  + 
  ## Add custom legend for dosage in greyscale
  geom_point(data = legend_dose, mapping = aes(x = x, y = y, fill = Label),
             inherit.aes = FALSE, shape=17, size = 0, alpha = 0,          # invisible
             show.legend = TRUE) +
  scale_fill_manual(values = c(legend_dose$Color), name = "# Bitter Taste Alleles ") + 
  guides(fill=guide_legend(override.aes = list(size=2, alpha=1, color=legend_dose$Color))) + 
  fig3_legend + theme(
    legend.direction = "horizontal",
    legend.position = c(0.5, 0.98),  # Moves legend to bottom-right inside plot
    legend.justification = c(0.5, 1),
    legend.key.size = unit(0.55, 'line'),
    legend.margin = margin(3,3,3,3, unit="pt"),
    legend.title = element_text(margin=margin(r=4, b=0, unit="pt")),
    legend.background = element_rect(fill = "white", color = "black", linewidth=0.15), 
    ) ; panelD

```

```{r}
ggarrange(
  ggarrange(panelA, panelB, ncol=1, align="v", labels=c("A","B")),
  ggarrange(panelC, panelD, ncol=1, heights=c(2,1.25), labels=c("C","D")), 
  ncol=2, widths=c(0.75,1.25), align="h"
) %>% ggsave(filename="../output/manuscript/plot_figure3_v2.pdf", height=5, width=8)


```

```{r}  
###################################
###  Run Sex-Stratified Models  ###
###################################

sex.l <- list(c("Female", "Male"))

## NO significanat sex interactions
summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num*sex+", adj_models[[1]])), data=analysis))
summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num*sex+", adj_models[[2]])), data=analysis))

## Sex-stratified associations with 0-2hr glucose
# BMI model
print_lm(exposure = "taste_diplos", outcome="glu2hr_mgdl", covariates = gsub("[+]sex","", adj_models[[1]]),
         label="Female", data=analysis %>% filter(sex=="Female"))
print_lm(exposure = "taste_diplos", outcome="glu2hr_mgdl", covariates = gsub("[+]sex","", adj_models[[1]]),
         label="Female", data=analysis %>% filter(sex=="Male"))

#summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("[+]sex","", adj_models[[1]]))), data=analysis %>% filter(sex=="Female")))
#summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("[+]sex","", adj_models[[1]]))), data=analysis %>% filter(sex=="Male")))

# Diet+Lifestyle adjusted model
summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("[+]sex","", adj_models[[4]]))), data=analysis %>% filter(sex=="Female")))
summary(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("[+]sex","", adj_models[[4]]))), data=analysis %>% filter(sex=="Male")))


## Run sex-stratified models, 
do.call(rbind.data.frame, lapply(1:3, function(g) {
  do.call(rbind.data.frame, lapply(sex.l, function(x) {
    if(glycemic_vars_mgdl.l[[g]]=="glu2hr_mgdl") {
      models.l.use<-gsub("sex","",models.sensitivity.nofast.l$Diet.Patterns) ; dat.use<-analysis %>% filter(fast_cat=="0to2hr")} else {
      models.l.use <- gsub("sex","",models.sensitivity.l$Diet.Patterns) ; dat.use <- analysis} ; return(
        do.call(rbind.data.frame, lapply(1:length(models.l.use), function(v) { 
          print_lm(exposure = "taste_diplos", outcome = glycemic_vars_mgdl.l[[g]], 
                   label.outcome = "temp" ,##paste0(x, ".", names(glycemic_vars_mgdl.l)[g]),
                   covariates = models.l.use, label="Diet.Patterns", data=dat.use %>% 
                     filter(sex == x), lm_trend = F) })) ) })) })) %>% 
  write.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplos_sexStrat_mDiet_mgdl_",tag,".csv"), row.names = F)


confint(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("sex", "", models.sensitivity.nofast.l$Diet.Patterns))), data=analysis %>% filter(sex=="Female")))
confint(lm(formula(paste0("glu2hr_mgdl~taste_diplos.num+", gsub("sex", "", models.sensitivity.nofast.l$Diet.Patterns))), data=analysis %>% filter(sex=="Male")))
```

 
bitter taste receptors & bitter food preference
```{r} 

bitter_prefs <- readRDS("../../diet-pgs-ses/data/processed/ukb_MA_analysis_allphenos.rda") %>%
  select(id, pref_bitter_food, pref_sweet_food, pref_salty_food, pref_fatty_food)

analysis <- analysis %>% left_join(bitter_prefs)

summary(lm(formula(paste0("pref_bitter_food~taste_diplos+", models.l$Base)), data=analysis))
summary(lm(formula(paste0("pref_bitter_food~as.factor(rs10772420_A)+", models.l$Base)), data=analysis))
 
```

