---
title: "sensitivity_summary"
output: html_document
date: "2024-09-05"
---

```{r setup, include=FALSE, echo=F}
knitr::opts_chunk$set(echo = F, error = F, warning = F, out.width = "110%", out.height = "50%")

lapply(c("tidyverse", "data.table", "paletteer", "ggpubr", #"gplots",
          "xtable", "knitr", "tinytex"),  library, character.only = TRUE)

source("../scripts/pantry.R", echo=F)

```


```{r}

###############################
##   Load data for EUR only  ##
###############################

# system arguments
EUR=c("European"="EUR")
tag="v6"
ANC="EUR"

# load postprocessed dataframe
postprocessed <- readRDS("../data/processed/ukb_postprocessed_EUR.rda") %>%
  mutate(N_sensitivity=ifelse(n_geno==1 & n_t2d_contrl==1 & n_covars==1 & n_fast==1 & n_fast_24hr==1,1,0))

analysis <- readRDS(paste0("../data/processed/ukb_analysis_", ANC, ".rda")) %>%
  mutate(n_24hr.lab = ifelse(n_24hr==1, "Has 24HR", "No 24HR")) %>%
  mutate(n_24hr_plaus.lab = factor(ifelse(n_24hr_plaus, "Plausible 24HR", "No 24HR"), levels=c("Plausible 24HR", "No 24HR")))

#load dietPCs
dietPCs.dat <- readRDS(paste0("../data/processed/diet/ukb_EUR_dietPCs.rda"))
dietPC_emm <- read.csv(paste0("../data/processed/descr_tab_emm_dietPC_diplos_sexageac_",tag,".csv"))

#Set variables for workflow
n24HR.l <- c("Has 24HR", "No 24HR")
n24HR_plaus.l <- c("Plausible 24HR", "No 24HR")

dim(analysis) #N=241378

```

***

# Sensitivity Analyses 
```{r}
n_sensitivity <- c(n_geno = "Genotyping, n(%)", n_t2d_contrl="T2D controls, n(%)" ,
  n_covars="Complete covariate data, n(%)", n_fast="Complete fasting data, n(%)",
  n_fast_24hr="Complete fasting data for <24hr, n(%)", n_24hr= "Available 24HR data, n (%)",
  N_sensitivity="Total sample for sensitivity analysis, N (%)") ; print_summary_table(
    data=postprocessed, var_strata = F, vars_to_summarize = n_sensitivity,
    factor_vars = c(names(n_sensitivity)), p_print=F) %>% 
  kable(caption="Sensitivity analysis sample among all EUR participants")
```

```{r}
m1_base = "age+sex+gPC1+gPC2+gPC3+gPC4+gPC5+gPC6+gPC7+gPC8+gPC9+gPC10+ac.f+fast_cat"
m2_bmi = paste0(m1_base, "+bmi")
m3_lifestyle = paste0(m2_bmi, "+smoke_level.lab+physact_level.lab+alch_freq.lab")
m4_diet = paste0(m3_lifestyle, "+dietPC3+dietPC4+dietPC6+dietPC11+dietPC12+dietPC13+dietPC15+dietPC16+dietPC17")
models.l = list("Base"= m1_base, "BMI"=m2_bmi, "Lifestyle"=m3_lifestyle, "Diet.Patterns"=m4_diet)

# sensitivity models
ms_ses = paste0(m4_diet, "+income_level.lab+educ_isced.lab")
ms_diet_all = paste0(m3_lifestyle, "+", paste0("dietPC",1:24,collapse = "+"))
ms_f2c = paste0(m4_diet, "+FIB2CHO")
ms_f2c_ses = paste0(m4_diet, "+FIB2CHO+income_level.lab+educ_isced.lab")
ms_tech = paste0(m4_diet, "+gluB_aliquot.0+gluB_assay_date.0+dilution")
ms_covars = gsub("alch_freq.lab", "alch_drinks_per_week", gsub("physact_level.lab", "physact_met_excess", m4_diet))

models.l = list("Base"= m1_base, "BMI"=m2_bmi, "Lifestyle"=m3_lifestyle, "Diet.Patterns"=m4_diet)
models.nofast.l <- as.list(gsub("[+]fast_cat", "", models.l)) ; names(models.nofast.l) <- names(models.l)
models.sensitivity.l = list("Diet.Patterns"=m4_diet, "SES"=ms_ses, "All.Diet.PCs"=ms_diet_all, "Fib2Carb"=ms_f2c, "Fib2Carb+SES"=ms_f2c_ses, 
                            "Technical"=ms_tech, "Alt.Covariates"=ms_covars)
models.sensitivity.nofast.l <- as.list(gsub("[+]fast_cat", "", models.sensitivity.l)) ; names(models.sensitivity.nofast.l) <- names(models.sensitivity.l)
```

## Sensitivy Models

### Covariate sensitivty/precision
Are covariates capturing expected/intended variance?
```{r}
# associations of covariates with glycemic outcomes
read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_lifestyle_ses_mBMI_mgdl_",tag,".csv")) %>%
  kable(caption="Associations of lifestyle/ses covariates with glycemic measures")
read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_covars_dietPCs_mgdl_",tag,".csv")) %>%
  kable(caption="Associations of dietPC covariates with glycemic measures")

read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_lifestyle_ses_cont_mBMI_mgdl_",tag,".csv")) %>%
  kable(caption="Associations of lifestyle/ses *more precise* covariates with glycemic measures")

# sensitivity models with alternative covariate coding
make_pretty_lm( read.csv(paste0("../data/processed/sens_tab_lm_all_tastediplos_mAltCovars_mDietAltCovars_",tag,".csv")) ) %>%
  kable(caption="Sensitivity models with *more precise* covariates")
```

```{r, include=F}
dpc1_median=quantile(analysis$dietPC1, probs=seq(0,1,0.5), include.lowest=F)[-1]
dpc1_tertiles=quantile(analysis$dietPC1, probs=seq(0,1,0.33), include.lowest=F)[-1]

## Based on dietPC1 (generally healthy/unhealthy indicators)
analysis <- analysis %>% mutate(
  dpc1_med = factor(ifelse(dietPC1 >= dpc1_median[1],"Above","Below"), levels=c("Below", "Above")),
  dpc1_tert = factor(case_when(dietPC1<dpc1_tertiles[1] ~ "Low",
                          dietPC1>=dpc1_tertiles[1] & dietPC1<dpc1_tertiles[2] ~ "Moderate",
                          dietPC1>dpc1_tertiles[2] ~ "High"), levels=c("Low", "Moderate", "High"))
)

# models stratified by 'healthy/less healthy' diets based on dietPC1
read.csv("../data/processed/sens_tab_descr_dPC1median_v6.csv") %>% kable()
ggarrange(analysis %>% ggplot(aes(x=glu, fill=dpc1_med)) + geom_histogram(bins=100, alpha=0.6),
          analysis %>% ggplot(aes(x=glu2hr, fill=dpc1_med)) + geom_histogram(bins=100, alpha=0.6),
          analysis %>% ggplot(aes(x=hba1c, fill=dpc1_med)) + geom_histogram(bins=100, alpha=0.6), nrow=1, common.legend = T)

make_pretty_lm( read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplos_dPC1median_mSES_mgdl_",tag,".csv")) ) %>%
  kable(caption="Sensitivity models below/above dPC1 median ")

# models stratified by 'lower healthy/less healthy' diets based on dietPC1
read.csv("../data/processed/sens_tab_descr_dPC1tertile_v6.csv") %>% kable()
ggarrange(analysis %>% ggplot(aes(x=glu, fill=dpc1_tert)) + geom_histogram(bins=100, alpha=0.6),
          analysis %>% ggplot(aes(x=glu2hr, fill=dpc1_tert)) + geom_histogram(bins=100, alpha=0.6),
          analysis %>% ggplot(aes(x=hba1c, fill=dpc1_tert)) + geom_histogram(bins=100, alpha=0.6), ncol=3, common.legend = T)

make_pretty_lm( read.csv(paste0("../data/processed/sens_tab_lm_all_tastediplos_dPC1tertile_",tag,".csv")) ) %>%
  kable(caption="Sensitivity models with *more precise* covariates")
```

Re-run with continuous fasting hours
```{r}
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplos_mFastHrs_mgdl_",tag,".csv"))) %>% 
  kable(caption="Adjusting for continuous fasting hours")

## Plot of model fit with categorical vs continuous fasting variable
models.fasthrs.l <- lapply(models.l, function(m) gsub("fast_cat", "fasting_hrs", m))
#adjR2_fast <- do.call(rbind, lapply(1:4, function(m) {
#  cbind(cat=summary(lm(formula(paste0("glu_mgdl~taste_diplos+", models.l[[m]])), data=analysis))$adj.r.squared,
#        cont=summary(lm(formula(paste0("glu_mgdl~taste_diplos+", models.fasthrs.l[[m]])),data=analysis))$adj.r.squared
#  )})) ; adjR2_fast %>% kable(caption="Adjusted R2 in primary models with categorical or continuous fasting")
```

**
### Additionally adjust for SES
```{r}
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_rg_2hg_tastediplos_mSES_mgdl_",tag,".csv"))) %>% 
  kable(caption="Associations of TAS2R38 with Glycemic Measures, additionally adjusting for SES")
  
```

**
### Additionally adjust for ALL 24 diet PCs
```{r}
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_rg_2hg_tastediplos_mAllDietPC_mgdl_",tag,".csv"))) %>% 
  kable(caption="Associations of TAS2R38 with Glycemic Measures, Adjusting for All 24 Diet PCs")

make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplos_num_mAllDietPC_mgdl_",tag,".csv"))) %>% 
  kable(caption="Associations of continuous TAS2R38 with Glycemic Measures, Adjusting for All 24 Diet PCs")

```


## II. Additionally adjust for fiber-to-carbohydrate ratio 

Analyses were performed in a subsample of participants with who **completed 24HRs** and had plausible values for total energy intake (kcal/day) derived using from self-reported intakes. 

Plausible energy intake was defined as: 
*for M: 600- 4880 kcal/day (~20 MJ)
*for F: 600- 4300 kcal/day (~18 MJ)

```{r, fig.width=12, include=F}
#### Bar plots of RG by 24HR data & fasting time 
# RG --------------
analysis %>%
  group_by(n_24hr_plaus.lab) %>%
  reframe(mean=mean(glu, na.rm=T),
          sd=sd(glu, na.rm=T)) %>%
  ggplot(aes(x=n_24hr_plaus.lab, y=mean-4, ymin=mean-4-sd, ymax=mean-4+sd, fill=as.factor(n_24hr_plaus.lab))) +
  geom_bar(stat = "identity", position=position_dodge(0.8)) +
  geom_errorbar(position=position_dodge(0.8), width=0.5) +
  scale_fill_manual(values=palettes$plaus_24hr, name="Plausible_Kcal") +  
  scale_y_continuous(breaks=seq(0,2.5,0.5), labels=seq(0,2.5,0.5)+4) + 
  ylab("Random Glucose, mmol/L") + xlab("Self-reported fasting time, hrs")

# Glu x fasting (1-24) -----------
analysis %>%
  select(glu, fasting_hrs, n_24hr_plaus.lab) %>%
  group_by(fasting_hrs, n_24hr_plaus.lab) %>%
  reframe(mean=mean(glu, na.rm=T),
          sd=sd(glu, na.rm=T)) %>%
  ggplot(aes(x=fasting_hrs, y=mean-4, ymin=mean-4-sd, ymax=mean-4+sd, fill=as.factor(n_24hr_plaus.lab))) +
  geom_bar(stat = "identity", position=position_dodge(0.8)) +
  geom_errorbar(position=position_dodge(0.8), width=0.5) +
  scale_fill_manual(values=palettes$plaus_24hr, name="Has 24HR data") +  
  scale_y_continuous(breaks=seq(0,2.5,0.5), labels=seq(0,2.5,0.5)+4) + 
  scale_x_continuous(breaks=seq(0,23,1), labels = seq(1,24,1)) +
  ylab("Random Glucose, mmol/L") + xlab("Self-reported fasting time, hrs")
```

**Stratified models by 24HR data**
```{r}
#descriptive tables
read.csv(paste0("../data/processed/sens_descr_tab_24HRplaus_",tag,".csv")) %>% 
  kable(caption="Decription of subsample with available and plausible 24HR data")
read.csv(paste0("../data/processed/sens_descr_tab_24HR_center_",tag,".csv")) %>%
  kable(caption="Assessment Center x 24HR")

# Add adjustment for f2b and f2c+ses
make_pretty_lm( read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplo_mF2C_mgdl_",tag,".csv")), digits=c(3,3)) %>%
  kable(caption="TAS2R38 diplotype with RG & 2hr glucose, adjusting for Fiber-to-Carbohydrate ratio (& SES) ")

make_pretty_lm( read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplo_num_mF2C_mgdl_",tag,".csv")), digits=c(3,3)) %>%
  kable(caption="Additive TAS2R38 diplotype with RG & 2hr glucose, adjusting for Fiber-to-Carbohydrate ratio (& SES) ")

# for comparison, sample without 24hr data
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_rg_2hg_tastediplos_mNoF2C_mgdl_",tag,".csv")), digits=c(3,3)) %>%
  kable(caption="TAS2R38 diplotype with RG & 2hr glucose, among sample without F2C data")
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_rg_2hg_tastediplos_num_mNoF2C_mgdl_",tag,".csv")), digits=c(3,3)) %>%
  kable(caption="Additive TAS2R38 diplotype with RG & 2hr glucose, among sample without F2C data")

```

### Sex-stratified models
```{r}
make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_tastediplos_sexStrat_mDiet_mgdl_",tag,".csv"))) %>%
  kable(caption="Sex-stratified associations")
```

## II. Examine bitter SNPs (caffeine & quinine) as negative controls

Alternative genetic instruments were used for bitter taste as a negative control

```{r}
snps_bitter_diet <- read.csv(paste0("../data/processed/sens_descr_tab_bitterDiet_bitterSNPs_",tag,".csv"))
make_pretty_lm(snps_bitter_diet %>% rename(outcome=Model, exposure=SNP) %>% mutate(model="Age,Sex,gPCs")) %>%
  kable(caption = "GLMs for other bitter SNPs with dietary habits & PCs")

snps_bitter_diet.mat <- as.matrix(
  snps_bitter_diet %>% dplyr::select(SNP, Diet_Trait=Model, Beta=beta) %>%
  pivot_wider(values_from=Beta, names_from=Diet_Trait) %>%
  dplyr::select(-SNP) %>% t()
) ; colnames(snps_bitter_diet.mat) <- unique(snps_bitter_diet$SNP)

#heatmap.2(t(snps_bitter_diet.mat),
#          col=palettes$hm_palette_rwb, scale = "none", 
#          linecol =  "black", Rowv=T, Colv=F,
#          density.info = "none", trace = "none", margins = c(6,7),
#          dendrogram = "none", #key="False",
#          cexRow = 0.75, cexCol = 0.75,
#          offsetRow=0, offsetCol = 0) 


```

#### Associations of other bitter SNPs with glycenic measures 
```{r}
bitter_snps <- c("rs713598_G", "rs1726866_G", "rs10246939_C", "rs10772420_A", "rs2597979_G")

#make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_allglycemic_bittersnps_mgdl_",tag,".csv"))) %>%
#  filter(Exposure %in% c("rs10772420_A", "rs2597979_G")) %>%
#  kable(caption = "Bitter SNPs & Glycemic Measures")

make_pretty_lm(read.csv(paste0("../data/processed/sens_tab_lm_gluXfast_bittersnps_mDiet_mgdl_",tag,".csv"))) %>%
  filter(Exposure %in% c("rs10772420_A", "rs2597979_G")) %>%
  kable(caption = "Bitter SNPs & Glucose by fasting time") 


```

#### Estimated marginal mean 0-2hr glucose for other SNPs
```{r, fig.width=10, fig.height=8, include=F}

emm_bittersnp_x_gluFast <- read.csv(paste0("../data/processed/sens_tab_emm_gluXfast_bittersnps_mDiet_mgdl_",tag,".csv"))
bitter_snps_cat <- paste0(bitter_snps, ".cat")

# Dot plot ---------------
p_emm_bittersnps_x_gluFast <- 
emm_bittersnp_x_gluFast %>% as.data.frame() %>%
  dplyr::mutate(fast = factor(fast, levels=c("6+hr", "5hr", "4hr", "3hr", "0to2hr"),
                       labels =c(">6 hrs", "5 hrs", "4 hrs", "3 hrs", "0-2 hrs")),
         Genotypes = factor(level, levels=c("0_0","0_1", "1_1"), labels=c("00", "01", "11")), .before=outcome) %>%
  mutate(SNP = rep(bitter_snps, each=15), .before=level) %>%
  mutate(Gene = c(rep("TAS2R38-PROP/PTC", 15*3), rep(c("TAS2R19-Quinine", "TAS2R14-Caffeine"), each=15))) %>%
  mutate(color = paste0(Gene, ".", Genotypes)) %>%
  mutate(SNP_Taste = ifelse(SNP %in% c(bitter_snps[1:3]), "TAS2R38 Bitter Taste Receptor", "Other bitter taste perception")) %>%
  filter(fast == "0-2 hrs") %>%
  ggplot(aes(y=emmean-88.25, x=SNP, ymin=lowCI-88.25, ymax=upCI-88.25, color=color)) + 
  scale_y_continuous(limits=c(-0.5,2), breaks=seq(-0.5,2,0.5), labels = seq(-0.5,2,0.5)+88) +
  facet_grid(~Gene, space="free_x", scales="free_x") +
  #geom_hline(yintercept = seq(4.88,4.98,0.02)-4.88, linewidth=0.15, color="grey") +
  geom_point(position=position_dodge(0.55), size=2) + 
  geom_errorbar(position=position_dodge(0.55), linewidth=0.75, width=0.55) + 
  scale_color_manual(values=c(palettes$alleles_browns, palettes$alleles_yellows, palettes$alleles_greyblue), name="Genotype") + 
  ylab("Marginal Mean (95% CI)\nGlucose Level, mg/dL") + xlab("") +
  theme_bw() +
  theme(legend.position = "bottom",
        legend.title = element_text(face="bold", size=9),
        axis.title = element_text(size=12), 
        axis.text.x = element_text(size=10, color="black"), axis.text.y = element_text(size=10, color="black"),
        panel.grid.major.y = element_line(), 
        strip.background = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.background = element_rect(fill="#00000000"),
        strip.text = element_text(face="bold", size=10)
  ) ; p_emm_bittersnps_x_gluFast

p_emm_bittersnps_x_gluFast %>% ggsave(filename="../output/plot_emm_glu2hr_mgdl_bitterSNPs_mDiet_v6_wh.png", height=3.75, width=7)

```

```{r, include=F}

## Overall all fasting windows?
emm_bittersnp_x_gluFast %>% as.data.frame() %>%
  dplyr::mutate(fast = factor(fast, levels=c("6+hr", "5hr", "4hr", "3hr", "0to2hr"),
                       labels =c(">6 hrs", "5 hrs", "4 hrs", "3 hrs", "0-2 hrs")),
         Genotypes = factor(level, levels=c("0_0","0_1", "1_1"), labels=c("00", "01", "11")), .before=outcome) %>%
  mutate(SNP = rep(bitter_snps, each=15), .before=level) %>%
  mutate(Gene = c(rep("TAS2R38-PROP/PTC", 15*3), rep(c("TAS2R19-Quinine", "TAS2R14-Caffeine"), each=15))) %>%
  mutate(color = paste0(Gene, ".", Genotypes)) %>%
  mutate(SNP_Taste = ifelse(SNP %in% c(bitter_snps[1:3]), "TAS2R38 Bitter Taste Receptor", "Other bitter taste perception")) %>%
  #filter(fast == "0-2 hrs") %>%
  ggplot(aes(y=fast, x=emmean-88.25,xmin=lowCI-88.25, xmax=upCI-88.25, color=color, group=fast)) + 
  #scale_y_continuous(limits=c(0,1.5), breaks=seq(0,1.5,0.5), labels = seq(0,1.5,0.5)+88) +
  facet_grid(~Gene, space="free_x", scales="free_x") +
  #geom_hline(yintercept = seq(4.88,4.98,0.02)-4.88, linewidth=0.15, color="grey") +
  geom_point(position=position_dodge2(0.35)) + geom_errorbar(position=position_dodge2(), width=0.35) + 
  scale_color_manual(values=c(palettes$alleles_browns, palettes$alleles_yellows, palettes$alleles_greyblue), name="Genotype")


```

## Causal Panel plots 
```{r}
#negative control data
negcont <- read.csv(paste0("../data/processed/sens_tab_lm_gluXfast_bittersnps_mDiet_mgdl_",tag,".csv")) %>%
  filter(model %in% c("BMI", "Diet.Patterns") & fast == "0to2hr") %>%
  arrange(model) %>%
  mutate(Bitter_SNP = factor(exposure, levels=rev(bitter_snps)),
         Model = factor(model, levels=c("Diet.Patterns", "BMI")))

#independent replication data
repl <- read.csv("../data/meta_analysis_2hg.csv") %>%
  filter(Trait == "0to2hrG_adjBMI") %>%
  select(snp=SNP, beta=BETA, se=SE, n=N) %>%
  rbind.data.frame(c("rs10772420_A", 0.009, 0.019, 0.6293),
                   c("rs2597979_G", 0.018, 0.025, 0.471)) %>% 
  mutate(Gene = c(rep("TAS2R38 (PROP/PTC)",3), "TAS2R19 (quinine)", "TAS2R14 (caffeine)")) %>%
  mutate(across(c("beta", "se", "n"), ~as.numeric(.))) %>%
  mutate(Bitter_SNP = factor(c("rs713598_G", "rs1726866_G", "rs10246939_C", "rs10772420_A", "rs1868769_G*"), 
                             levels=c("rs1868769_G*", rev(bitter_snps)[-1])),
         Model = "BMI") %>%
  mutate(across(c("beta", "se"), ~.*18.018))

#figure theme
fig3_theme <- theme_bw() + 
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        axis.text = element_text(color="black"),
        legend.key.size = unit(0.5, "cm"),
        legend.margin = margin(0.5,0.5,0.5,0.5),
        legend.key.spacing = unit(0.1, "cm"),
        legend.title = element_text(face="bold"),
        legend.position = "right")
```

```{r}
# Replication in OGTT  
p_negcont <- negcont %>%
  mutate(Gene = rep(c(rep("TAS2R38 (PROP/PTC)",3), "TAS2R19 (quinine)", "TAS2R14 (caffeine)"), 2)) %>%
  ggplot(aes(x=beta, xmin=beta-1.96*se, xmax=beta+1.96*se, y=Bitter_SNP, color=Gene, shape=Model)) +
  geom_vline(xintercept = 0, color="black", linewidth=0.25) +
  geom_errorbarh(height=0.25, position=position_dodge(0.5), linewidth=0.25) + 
  geom_point(size=2, position=position_dodge(0.5)) + 
  scale_shape_manual(values=c(19,1)) + 
  scale_x_continuous(limits = c(-0.5, 0.5))  +
  xlab("Beta (95% CI)") + ylab("") +
  scale_color_manual(values=c(palettes$alleles_browns[3], palettes$alleles_yellows[3], palettes$alleles_greyblue[3]), name="Gene (tastant)") + 
  fig3_theme

p_repl <- repl %>%
  ggplot(aes(x=beta, xmin=beta-1.96*se, xmax=beta+1.96*se, y=Bitter_SNP, color=Gene, shape=Model)) +
  geom_vline(xintercept = 0, color="black", linewidth=0.25) +
  geom_point(size=2, position=position_dodge(0.5)) + 
  geom_errorbarh(height=0.25, position=position_dodge(0.5), linewidth=0.25) + 
  scale_shape_manual(values=1) + 
  scale_x_continuous(limits = c(-1.25, 1.25))  +
  xlab("Beta (95% CI)") + ylab("") +
  scale_color_manual(values=c(palettes$alleles_browns[3], palettes$alleles_yellows[3], palettes$alleles_greyblue[3]), name="Gene (tastant)") + 
  fig3_theme

ggarrange(p_negcont,"", p_repl, common.legend = T, widths = c(1,0,1), legend = "right", nrow=1) %>%
  ggsave(filename="../output/plot_panel_bradford.pdf", height=3, width=8.25)
```

```{r}
## EMMs
p_emm_negcont <- emm_bittersnp_x_gluFast %>% as.data.frame() %>%
  mutate(model="Diet") %>% filter(fast=="0to2hr") %>%
  mutate(Genotypes = factor(level, levels=c("0_0","0_1", "1_1"), labels=c("00", "01", "11")), .before=outcome) %>%
  mutate(SNP = factor(gsub(".cat", "", exposure), levels=bitter_snps)) %>%
  mutate(Gene_taste = c(rep("TAS2R38-PROP/PTC", 9), rep("TAS2R19-Quinine",3), rep("TAS2R14-Caffeine",3) )) %>%
  mutate(Gene = gsub("-.*", "", Gene_taste)) %>%
  mutate(color = paste0(Gene, " (", Genotypes, ")")) %>%
  ggplot(aes(y=emmean-88.25, x=SNP, ymin=lowCI-88.25, ymax=upCI-88.25, color=color)) + 
  scale_y_continuous(limits=c(-0.5,2), breaks=seq(-0.5,2,0.5), labels = seq(-0.5,2,0.5)+88) +
  geom_point(position=position_dodge(0.55), size=2) + 
  geom_errorbar(position=position_dodge(0.55), linewidth=0.5, width=0.25) + 
  scale_color_manual(values=c(palettes$alleles_browns, palettes$alleles_yellows, palettes$alleles_greyblue), name="Genotype") + 
  ylab("Blood glucose (mg/dL)") + xlab("") +
  theme_bw() +
  theme(panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        axis.text = element_text(color="black"),
        legend.key.size = unit(0.45, "cm"),
        legend.text = element_text(color="black", size=8))

p_emm_negcont %>% ggsave(filename="../output/plot_panel_emm_bradford.pdf", height=2.5, width=8)

```


## Combined specificity and replication results
#### Take-aways

For Plausible kcal:

* No significant Diplotye x Plausible_kcal interaction
* In models with an interaction term, no significant diplotype-glucose associations...
* Plausible_kcal is the strongest predictor of RG (associations become less significant as you adjust for lifestyle then diet patterns)

Sex

* No significant Diplotype x Sex interactions in any model
* In models with a sex interaction term, no significant independent diplotype-glucose association
* Sex is a significant predictor in Lifestyle & Diet Patterns model

Income

* NO significant diplotype x income interaction in any model
* in models with interaction, no significant diplotye-glucose association
* income 31-52 & 52-100 significant predictors of lower RG relative to <18k in base model
* Associations are strengthened in the lifestyle and diet patterns model
* 

Education

* Significant diplotype x education interaction (in some levels) in all models
* In models with interaction term, diplotype remains a significant predictor
* Interactions: College vs none of the above



